onnxruntime/core/providers/opencl/opencl_program_manager.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include "core/common/common.h"
+#include "core/framework/murmurhash3.h"
+
+#include "opencl_forward_decl.h"
+#include "opencl_utils.h"
+#include "opencl_execution_provider.h"
+
+#include <string_view>
+
+namespace onnxruntime {
+namespace opencl {
+
+using ProgramKey = std::array<uint32_t, 4>;
+using KernelKey = std::pair<cl_program, std::string>;
+
+}  // namespace opencl
+}  // namespace onnxruntime
+
+template <>
+struct std::hash<onnxruntime::opencl::ProgramKey> {
+  size_t operator()(const onnxruntime::opencl::ProgramKey& key) const {
+    std::hash<uint32_t> h{};
+    auto v = 0;
+    v = onnxruntime::opencl::HashCombine(v, h(key[0]));
+    v = onnxruntime::opencl::HashCombine(v, h(key[1]));
+    v = onnxruntime::opencl::HashCombine(v, h(key[2]));
+    v = onnxruntime::opencl::HashCombine(v, h(key[3]));
+    return v;
+  }
+};
+
+template <>
+struct std::hash<onnxruntime::opencl::KernelKey> {
+  size_t operator()(const onnxruntime::opencl::KernelKey& kernel_key) const {
+    auto a = std::hash<size_t>{}(reinterpret_cast<size_t>(kernel_key.first));
+    auto b = std::hash<std::string>{}(kernel_key.second);
+    return onnxruntime::opencl::HashCombine(a, b);
+  }
+};
+
+namespace onnxruntime {
+namespace opencl {
+
+class OpenCLKernelHolder {
+ public:
+  explicit OpenCLKernelHolder(OpenCLProgramManager* mgr) : mgr_{mgr} {}
```

Comment:
nit: If mgr cannot be nullptr pass it by reference to advertise that. You can still store it internally as a pointer it needed (reference is preferred for clarity though).

Reply:
Done